File: string/[프로그래머스]Lv1. 81301. 숫자 문자열과 영단어/solution.py
'''
num_dic = {"zero":"0", "one":"1", "two":"2", "three":"3", "four":"4", "five":"5", "six":"6", "seven":"7", "eight":"8", "nine":"9"}

def solution(s):
    answer = s
    for key, value in num_dic.items():
        answer = answer.replace(key, value)
    return int(answer)
'''

def solution(s):
    answer = 0
    numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
    l = len(s)
    i = 0
    while i < l:
        if s[i] in numbers:
            add_num = int(s[i])
            i += 1
        elif s[i] == 'z':
            add_num = 0
            i += 4
        elif s[i] == 'o':
            add_num = 1
            i += 3
        elif s[i] == 't':
            if s[i + 1] == 'w':
                add_num = 2
                i += 3
            else:
                add_num = 3
                i += 5
        elif s[i] == 'f':
            if s[i + 1] == 'o':
                add_num = 4
            else:
                add_num = 5
            i += 4
        elif s[i] == 's':
            if s[i + 1] == 'i':
                add_num = 6
                i += 3
            else:
                add_num = 7
                i += 5
        elif s[i] == 'e':
            add_num = 8
            i += 5
        elif s[i] == 'n':
            add_num = 9
            i += 4
        answer = answer * 10 + add_num

    return answer

File: string/[프로그래머스]Lv1. 81301. 숫자 문자열과 영단어/solution.cpp
#include <vector>
#include <regex>

using namespace std; 

// regex 이용한 풀이 (속도 느림)
#include <string>
int solution(string s) {
    vector<string> v= {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
    for (int i = 0; i < 10; i++) {
        regex re(v[i]);
        // regex re = v[i] (x)
        // regex re = regex(v[i]) (o)
        s = regex_replace(s, re, to_string(i));
    }
    return stoi(s);
}

// 첫 풀이 (속도 빠름)
int solution(string s) {
    string answer = "";
    while (s.length() != 0) {
        if (s[0] == 'z') {
            answer += '0';
            s = s.substr(4);
        }
        else if (s[0] == 'o') {
            answer += '1';
            s = s.substr(3);
        }
        else if (s[0] == 't') {
            if (s[1] == 'w') {
                answer += '2';
                s = s.substr(3);
            }
            else {
                answer += '3';
                s = s.substr(5);
            }
        }
        else if (s[0] == 'f') {
            if (s[1] == 'o') {
                answer += '4';
                s = s.substr(4);
            }
            else {
                answer += '5';
                s = s.substr(4);
            }
        }
        else if (s[0] == 's') {
            if (s[1] == 'i') {
                answer += '6';
                s = s.substr(3);
            }
            else {
                answer += '7';
                s = s.substr(5);
            }
        }
        else if (s[0] == 'e') {
            answer += '8';
            s = s.substr(5);
        }
        else if (s[0] == 'n') {
            answer += '9';
            s = s.substr(4);
        }
        else {
            answer += s[0];
            s = s.substr(1);
        }
    }
    return stoi(answer);
}